application.ts

```diff
@@ -1,83 +1,153 @@
 
+// TODO: doc-strings with full examples
+
+// TODO: define ourselves
+import type { LevelName, Logger } from "@std/log";
+import { CoreDriverAdapter } from "./drivers/core_adapter.ts";
+import { type Driver, DriverStrategy, type Server } from "./drivers/driver.ts";
+import { ServerContext } from "./context.ts";
+import { buildContainer, registerClassMethods } from "./container.ts";
+import { buildControllerRoutes } from "./router.ts";
+import { ClassType } from "./utils.ts";
+
+/**
+ * A class which starts the API applications and allows one to register
+ * routes and capabilities to process inbound requests against.
+ */
 export class Application {
-  public listen(options?: ApplicationOptions): Promise<void> {
-    this.#server = Deno.serve({
-      port: options?.port ?? 8000,
-      hostname: options?.hostname ?? "0.0.0.0",
-      handler: this.#handler.bind(this),
-      onListen: this.#onListen.bind(this),
-    });
-    return this.finished;
-  }
+  readonly options: Readonly<ApplicationOptions>;
+  readonly #driver: Readonly<Driver>;
+  readonly #versions = new Map<string, ApplicationVersionOptions>();
 
-  /**
-   * A promise that resolves once the application finishes - eg. when aborted using the signal passed to ServeOptions.signal.
-   */
-  public get finished(): Promise<void> {
-    this.#assertStarted(this.#server);
-    return this.#server.finished;
-  }
+  public readonly ctx: Readonly<ServerContext>;
 
-  /**
-   * Make the application block the event loop from finishing.
-   *
-   * Note: the application blocks the event loop from finishing by default. This method is only meaningful after .unref() is called.
-   */
-  public ref(): void {
-    this.#assertStarted(this.#server);
-    return this.#server.ref();
-  }
+  public readonly log: Readonly<Logger>;
 
-  /**
-   * @returns
-   */
-  public unref(): void {
-    this.#assertStarted(this.#server);
-    return this.#server.unref();
-  }
+  constructor(options?: ApplicationOptions) {
+    // Driver
+    const defaultDriver = DriverStrategy.Core;
+    const driver = options?.driver ?? defaultDriver;
 
-  #server?: Deno.HttpServer;
+    // Log level
+    const defaultLogLevel: LevelName = "DEBUG";
+    const logLevel: LevelName = options?.logLevel ?? defaultLogLevel;
 
-  #assertStarted(
-    server: Deno.HttpServer | undefined,
-  ): asserts server is Deno.HttpServer {
-    assert(
-      server,
-      "Application is not yet started. Please start by calling: .serve()",
-    );
+    // Setup
+    const defaultOptions: Required<ApplicationOptions> = {
+      driver: defaultDriver,
+      logLevel: defaultLogLevel,
+    };
+    this.options = { ...defaultOptions, ...options };
+    this.ctx = new ServerContext(logLevel);
+    this.log = this.ctx.log;
+    switch (driver) {
+      case DriverStrategy.Core: {
+        this.#driver = new CoreDriverAdapter(this.ctx);
+        break;
+      }
+      default: {
+        this.#driver = driver;
```

Comment:
I like throwing errors when the wrong enums are used, we can say `unsupported` here

Reply:
Ah I was thinking this was where users could supply a custom driver that isn't in the enum? Maybe I can handle this in some separate logic! E.g.

```ts
if (typeof driver === 'function') {this.#driver=driver}

switch () {
  // ....
  default: {
     assertNever()
  }
}
```

wdyt?

Reply:
I agree.